sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala

```diff
@@ -165,6 +170,23 @@ object EliminateOuterJoin extends Rule[LogicalPlan] with PredicateHelper {
     case f @ Filter(condition, j @ Join(_, _, RightOuter | LeftOuter | FullOuter, _, _)) =>
       val newJoinType = buildNewJoinType(f, j)
       if (j.joinType == newJoinType) f else Filter(condition, j.copy(joinType = newJoinType))
+
+    case a @ Aggregate(_, _, join @ Join(left, _, LeftOuter, _, _))
+        if a.isDistinct && a.references.subsetOf(AttributeSet(left.output)) &&
+          !canPlanAsBroadcastHashJoin(join, conf) =>
```

Comment:
Sorry I missed this. It's outer join and it will never reduce data volume of the left side. So we can always remove the join, no matter it's broadcast or not.

Reply:
The result may be incorrect if always remove the join. For example:
```
0: jdbc:hive2://hdc49-mcc10-01-0510-2005-006-> create table test11.t1 using parquet as select id % 3 as a, id as b from range(10);
+---------+--+
| Result  |
+---------+--+
+---------+--+
No rows selected (1.611 seconds)
0: jdbc:hive2://hdc49-mcc10-01-0510-2005-006-> create table test11.t2 using parquet as select id % 3 as x, id as y from range(5);
+---------+--+
| Result  |
+---------+--+
+---------+--+
No rows selected (1.043 seconds)
0: jdbc:hive2://hdc49-mcc10-01-0510-2005-006-> select t1.a from t1 left join t2 on a = x;
+----+--+
| a  |
+----+--+
| 0  |
| 0  |
| 1  |
| 1  |
| 0  |
| 0  |
| 1  |
| 1  |
| 2  |
| 0  |
| 0  |
| 1  |
| 1  |
| 2  |
| 0  |
| 0  |
| 2  |
+----+--+
17 rows selected (1.409 seconds)
```

Reply:
The aggregate should still be there. I mean we can remove this `canPlanAsBroadcastHashJoin` check

Reply:
https://github.com/apache/spark/pull/32744